### PR TITLE
[th/wireguard-pt3 (v2)]

### PR DESCRIPTION
**Pull request by thom311**
This resurrects the work from https://github.com/NetworkManager/NetworkManager/pull/281 and is based on https://github.com/NetworkManager/NetworkManager/pull/293.

I think this is ready to merged. A few things are still missing. They are documented in `TODO` comments in `src/devices/nm-device-wireguard.c`. But these things should not block this branch. And I don't even think they would block a `1.16.0` release (with this branch in).

Please test and review. Thanks

### CODE REVIEW

File: src/devices/nm-device.c [hunk 1]
```diff
@@ -2044,9 +2044,14 @@ nm_device_get_route_metric_default (NMDeviceType device_type)
 	 */
 
 	switch (device_type) {
-	/* 50 is also used for VPN plugins (NM_VPN_ROUTE_METRIC_DEFAULT) */
+
+	/* 50 is also used for VPN plugins (NM_VPN_ROUTE_METRIC_DEFAULT).
+	 *
+	 * Note that returning 50 from this function, means that this device-type is
```
Comment (bengal) [inline src/devices/nm-device.c]: The comma after 'function' sounds bad. Also 'tread' -> 'treat' in commit subject.

File: shared/nm-utils/nm-macros-internal.h [hunk 2]
```diff
@@ -866,6 +866,17 @@ fcn (void) \
 #define NM_STR_HAS_PREFIX(str, prefix) \
 	(strncmp ((str), ""prefix"", NM_STRLEN (prefix)) == 0)
 
+#define NM_STR_HAS_SUFFIX(str, prefix) \
```
Comment (bengal) [inline shared/nm-utils/nm-macros-internal.h]: prefix -> suffix

File: shared/nm-utils/nm-macros-internal.h [hunk 3]
```diff
@@ -866,6 +866,17 @@ fcn (void) \
...
+	({ \
+		const char *_str = (str); \
+		gsize _l = strlen (_str); \
+		\
+		(   (_l >= NM_STRLEN (prefix)) \
+		 && (0 == memcmp (&_str[_l - NM_STRLEN (prefix)], \
```
Comment (bengal) [inline shared/nm-utils/nm-macros-internal.h]: Why is 0 on the left side? It seems unusual for our coding style.

Comment (thom311) [reply shared/nm-utils/nm-macros-internal.h]: the reason was, that the `== 0` at the end is easy to miss in
```
»···»···(   (_l >= NM_STRLEN (suffix)) \
»···»··· && (memcmp (&_str[_l - NM_STRLEN (suffix)], \
»···»···             ""suffix"", \
»···»···             NM_STRLEN (suffix)) == 0)); \
```

But I agree. Changed.

**Comment by thom311** [comment]
Change in binary size when building with
```
    $ ./contrib/fedora/rpm/build_clean.sh -g -w crypto_gnutls \
                 -W debug -w iwd -W test -w libnm_glib
```
on x64_86, Fedora 29

commit 9a4cd1efd1228c21c552546da4900e2ad3d4b92c (before) vs 489a6abec3ef33708c75530be1127db2aa1e5599 (after):
```
[before] 3360928 usr/sbin/NetworkManager
          +55504 (+ 1.65 %)
[after]  3416432 /usr/sbin/NetworkManager

[before] 1781456 /usr/lib64/libnm.so.0.1.0
         + 37584 (+ 2.11%)
[after]  1819040 /usr/lib64/libnm.so.0.1.0

[before]  896664 /usr/bin/nmcli
          + 8256 (+ 0.92%)
[after]   904920 /usr/bin/nmcli
```

File: libnm-core/nm-keyfile-utils.h [hunk 4]
```diff
@@ -25,7 +25,8 @@
 #error Cannot use this header.
 #endif
 
-#define NM_KEYFILE_GROUP_VPN_SECRETS "vpn-secrets"
+#define NM_KEYFILE_GROUP_VPN_SECRETS          "vpn-secrets"
+#define NM_KEYFILE_GROUPPREFIX_WIREGUARD_PEER "wireguard-peer."
```
Comment (fgiudici) [inline libnm-core/nm-keyfile-utils.h]: is the "." wanted in _wiregard-peer._ ?

File: src/devices/nm-device-wireguard.c [hunk 5]
```diff
@@ -26,30 +26,113 @@
...
+/* TODO: ensure externally-managed works. Both after start of NM and
+ *   when adding a wg link with NM running. */
+
+/* TODO: activate profile with peer preshared-key-flags=2. On first activation, the secret is
+ *   requested (good). Enter it and connect. Reactivate the profile, now there is no password
+ *   prompt, as the secret is cached (good??). */
+
+/* TODO: unlike for other VPNs, we don't inject a direct route to the peers. That means,
+ *   you might get a routing sceneraio where the peer (VPN server) is reachable via the VPN.
+ *   How we handle adding routes to external gateway for other peers, has severe issus
```
Comment (fgiudici) [inline src/devices/nm-device-wireguard.c]: typo: _has severe issu**e**s_

File: NEWS [hunk 6]
```diff
@@ -14,6 +14,9 @@ USE AT YOUR OWN RISK.  NOT RECOMMENDED FOR PRODUCTION USE!
 * Use a new type of secret-keys that combines the secret value with /etc/machine-id.
   This way when cloning a VM it suffices to change machine-id to generate different
   addresses.
+* Add support for WireGuard VPN tunnels to NetworkManager. D-Bus API and libnm
+  supports all options. nmcli supports creating and managing WireGuard profiles,
```
Comment (fgiudici) [inline NEWS]: typo: support-**s**- all options

File: libnm-core/nm-keyfile-utils.h (hunk 4 above)
Comment (thom311) [reply libnm-core/nm-keyfile-utils.h]: yes, it is. The prefix includes `'.'`.

**Comment by thom311** [comment]
Thanks @fgiudici. Addresses your comments and repushed.

**Comment by lkundrak** [comment]
LGTM

**Comment by thom311** [comment]
Thanks!! merged.

**Pull request merged**